### PR TITLE
Compatibility with Sunset Harbor

### PR DESCRIPTION
**Pull request by krzychu124**
Compatibility with Sunset Harbor

- fixed ` PathUnits.m_vehicleTypes` error,
- added Trolleybuses AI,
- fixed priorities to cover Trolleybuses
- bumped version

### CODE REVIEW

**Comment by krzychu124** [comment]
ops... forgot to update CI...

File: TLM/TMPE.API/Traffic/Enums/ExtVehicleType.cs [hunk 1]
```diff
@@ -27,15 +27,16 @@ public enum ExtVehicleType {
         PassengerFerry = 1 << 16,
         PassengerBlimp = 1 << 17,
         CargoPlane = 1 << 18,
+        Trolleybus = 1 << 20,
```
Comment (aubergine10) [inline TLM/TMPE.API/Traffic/Enums/ExtVehicleType.cs]: Why 20 and not 19?

Comment (krzychu124) [reply TLM/TMPE.API/Traffic/Enums/ExtVehicleType.cs]: Damn... I don't know what will break when I fix that value 😕

Comment (krzychu124) [reply TLM/TMPE.API/Traffic/Enums/ExtVehicleType.cs]: Anyways, good catch, I will test what happen, but I need to create new game for that

Comment (aubergine10) [reply TLM/TMPE.API/Traffic/Enums/ExtVehicleType.cs]: Does that info get stored in save?

Comment (krzychu124) [reply TLM/TMPE.API/Traffic/Enums/ExtVehicleType.cs]: I must check, if not I will update it

**Comment by kianzarrin** [comment]
I compared the code to other custom AI. 
There is a ton of dormant code in other AIs.
there is a lot of repeated code.

This is  just a comment. no need to make any changes in this PR.

**Comment by kianzarrin** [comment]
in the decompiler I noticed there is a line like this:
```
RoadBaseAI.ClickNodeButton() 
if ((info.m_vehicleTypes & (VehicleInfo.VehicleType.Car | VehicleInfo.VehicleType.Tram | 
VehicleInfo.VehicleType.Trolleybus)) != VehicleInfo.VehicleType.None)
```
I compared it to `CustomRoadBaseAI` where we have https://github.com/CitiesSkylinesMods/TMPE/blob/89f622a609fb5069bdad1256ca70f78b4619f11f/TLM/TLM/Custom/AI/CustomRoadBaseAI.cs#L42

So you might want to add trolly bus in `CustomClickNodeButton` too. Not sure if it is important or if the code is even used or not. from the name I get the impression the UI could be in trouble.

**Comment by kianzarrin** [comment]
What is the deal with StockPathFind.cs? its not even compiled. is it only for information? If that is the case it is now outdated. maybe delete it or put a comment that it is outdated ? preferably with an issue number.

speaking of the devil are there any issues tracking redundant AI/pathfind code? if there is its worth referencing the issue in the code
// TODO [issue #123] this code is redundant/outdated

**Comment by kianzarrin** [comment]
its worth to use `ExtVehicleManager.VEHICLE_TYPES` in 
https://github.com/CitiesSkylinesMods/TMPE/blob/89f622a609fb5069bdad1256ca70f78b4619f11f/TLM/TLM/Manager/Impl/ExtSegmentManager.cs#L112
and 
https://github.com/CitiesSkylinesMods/TMPE/blob/89f622a609fb5069bdad1256ca70f78b4619f11f/TLM/TLM/Manager/Impl/ExtSegmentEndManager.cs#L323

`ExtVehicleManager.VEHICLE_TYPES` will need `trollybus`

File: TLM/TLM/Custom/AI/CustomTrolleybusAI.cs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+            args.startPosB = startPosB;
+            args.endPosA = endPosA;
+            args.endPosB = endPosB;
+            args.vehiclePosition = default;
+            args.laneTypes = NetInfo.LaneType.Vehicle;
+            args.vehicleTypes = info.m_vehicleType;
+            args.maxLength = 20000f;
+            args.isHeavyVehicle = IsHeavyVehicle();
+            args.hasCombustionEngine = CombustionEngine();
+            ;
```
[the remark this answers is not in the document]
Comment (krzychu124) [reply TLM/TLM/Custom/AI/CustomTrolleybusAI.cs]: yeah, I don't know why it's there.

**Comment by aubergine10** [comment]
> What is the deal with StockPathFind.cs?

Each time there is game update we decompile the stock pathfinder and diff it to the StockPathFind.cs to see what's changed.

**Comment by aubergine10** [comment]
> I looked at several other places in the code and I can see that it works for as long as there will be no trolly-only custom road assets.

There are definitely trolly only roads being worked on.

**Comment by kianzarrin** [comment]
> @aubergine10  There are definitely trolly only roads being worked on.

@krzychu124 Please search for tram everywhere except for `TramBaseAI` and add trolley-bus where-ever you see tram (and deem it to be necessary) . There are not too many places so it will be easy.

Lane connector also does not work for trolley only roads.

**Comment by krzychu124** [comment]
@kianzarrin ok will do, also I've just noticed that nullpointer fix cause some error on second load, so possibly breaks hot-loading...

**Comment by aubergine10** [comment]
I'll ask around the creator communities to get clearer picture of what they're working on and rough ETA on when it might land in workshop.

**Comment by krzychu124** [comment]
I need to figure out what is wrong with PathManager and it will be ready for merge. 
Maybe it's worth bump version number to 11.2.1? What do you think?

**Comment by aubergine10** [comment]
Yup, we can refer to the PR in both sets of release notes, and just have one release for 11.2.1

**Comment by krzychu124** [comment]
I think NullPointerException was fixed. 
Already tested cases:
- enter + exit Map/Asset Editor,
- hotreload while in main menu,
- hotreload from in-game,
-  reload savegame being in-game,
- "back-to-menu and load different savegame" scenario

I didn't noticed any issues or errors. Would be good if someone could confirm these results.
 @aubergine10 Is there any pending issue to solve with regards to Sunset Harbor (or something simple that we can fit in update)?

**Comment by kianzarrin** [comment]
> There is a ton of dormant code in other AIs.
there is a lot of repeated code.

This has not been addressed but I don't consider it as part of this review so I approved.

**Comment by aubergine10** [comment]
@kianzarrin  can you create issue to track that domant code so we remember to sort it later?

**Comment by krzychu124** [comment]
@kianzarrin I'm not sure what do you mean by a lot of repeated/dormant code, but we will fix it in next PR. Thanks for review 😉

**Comment by kianzarrin** [comment]
@krzychu124 dormant ...  I made a mistake. I don't understand how redirection framework works.

repeated ... yes. for example trolley bus custom AI looks like a copy of tram custom AI with a couple of minor modifications. Ideally repeated code should be shared.

**Comment by krzychu124** [comment]
> Ideally repeated code should be shared.

No, because you will have hard time to spot a difference when they(CO) change anything

**Pull request merged**